File: backend/app/materials.py
import os
from functools import lru_cache
from supabase import create_client, Client


def _client() -> Client:
    url = os.getenv("SUPABASE_URL")
    key = os.getenv("SUPABASE_SERVICE_ROLE_KEY")
    if not url or not key:
        raise RuntimeError("SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY are not set — check your .env file.")
    return create_client(url, key)


@lru_cache(maxsize=1)
def load_catalog() -> dict[str, dict]:
    data = _client().table("materials").select("item_id, item_name, unit, price, category").execute()
    return {row["item_id"]: row for row in data.data}

File: src/lib/materials.ts
import { supabase } from "./supabase";

export type Material = {
  id: string;
  item_id: string;
  item_name: string;
  unit: string | null;
  price: number;
  category: string;
};

export async function fetchMaterials(): Promise<Material[]> {
  const { data, error } = await supabase
    .from("materials")
    .select("id, item_id, item_name, unit, price, category")
    .order("category", { ascending: true });

  if (error) {
    throw new Error(`Failed to load materials: ${error.message}`);
  }

  return data ?? [];
}

export async function fetchMaterialsByItemId(): Promise<Record<string, Material>> {
  const list = await fetchMaterials();
  return Object.fromEntries(list.map((m) => [m.item_id, m]));
}